### ISSUE TITLE
Resolution contains version range

### ISSUE THREAD

**Issue by eed3si9n** [coursier/coursier#2009]
## steps

Resolve `com.nimbusds:nimbus-jose-jwt:4.41.1`

## problem

`Resolution#dependencyArtifacts` returns `Dependency` value with version range inside.

## expectation

As a dependency resolver used in various tooling, my expectation is that Coursier would return the metadata of the final resolution in `Resolution`, not just the URLs but also the `Dependency` information with concrete version number that it settled on.

## note

For now I'm working around this using this function

```scala
  def actualDependency(d0: Dependency, projectCache: ProjectCache): Dependency =
    projectCache.get(d0.moduleVersion) match {
      case Some((_, p)) => d0.withVersion(p.actualVersion)
      case _            => d0
    }
```

but I don't think we should have to do this.

**Comment by alexarchambault** [comment]
I agree the concrete version would be better than the interval for that.

Do you want to fix it? The version you're looking for is available in [`reconciledVersions`](https://github.com/coursier/coursier/blob/bbc251e2b5be3dca415cbc335ff9c848256c7438/modules/core/shared/src/main/scala/coursier/core/Resolution.scala#L826-L830), which should be changed so that it uses `Project.actualVersion` rather than just `Project.version`.

Then, just in case, an override of `dependencyArtifacts` could be added, like
```scala
def dependencyArtifacts(classifiers: Option[Seq[Classifier]], classpathOrder: Boolean, reconciledVersions: Boolean) = ...
```
so that users can recover the former behavior if they want to, just in case. The other overrides of `dependencyArtifacts` would just pass `true` for `reconciledVersions`.

**Comment by alexarchambault** [comment]
(Just a note about naming: we have both [`retainedVersions`](https://github.com/coursier/coursier/blob/bbc251e2b5be3dca415cbc335ff9c848256c7438/modules/core/shared/src/main/scala/coursier/core/Resolution.scala#L832-L833) and `reconciledVersions`, and the definitions are reversed I think. The current `reconciledVersions` gives the "retained" concrete versions, and `retainedVersions` gives the versions right after the "version reconciliation" phase. This could be fixed in `2.1.x` I guess, it might break downstream things otherwise...)

**Comment by eed3si9n** [comment]
> Do you want to fix it?

I'll give it a shot.

**Comment by eed3si9n** [comment]
Here's my PR for this - https://github.com/coursier/coursier/pull/2020
